src/Parallel/Pin.cpp

```diff
@@ -74,9 +80,42 @@ cpu_set_t seissol::parallel::Pinning::getWorkerUnionMask() const {
 }
 
 cpu_set_t seissol::parallel::Pinning::getFreeCPUsMask() const {
-  // Affinity mask of the pthreads -> all free cores
+  const auto nodeOpenMpMask = getNodeMask();
+
   cpu_set_t freeMask{};
-  CPU_XOR(&freeMask, &processMask, &openmpMask);
+  CPU_ZERO(&freeMask);
+
+#ifdef USE_NUMA_AWARE_PINNING
+  // Find all numa nodes on which some OpenMP worker is pinned to
+  std::set<int> numaDomainsOfThisProcess{};
+  for (int cpu = 0; cpu < get_nprocs(); ++cpu) {
+    if (CPU_ISSET(cpu, &openmpMask)) {
+      numaDomainsOfThisProcess.insert(numa_node_of_cpu(cpu));
+    }
+  }
+
+  // Set free mask to all free threads which are on one of our numa nodes
+  for (int cpu = 0; cpu < get_nprocs(); ++cpu) {
+    const bool isFree = !CPU_ISSET(cpu, &nodeOpenMpMask);
+    if (isFree) {
+      const int numaNode = numa_node_of_cpu(cpu);
+      const bool isValidNumaNode = numaDomainsOfThisProcess.count(numaNode) != 0;
+      if (isValidNumaNode) {
+        CPU_SET(cpu, &freeMask);
+      }
+    }
+  }
```

Comment:
Does SeisSol crash if there is no free cpu? If yes, then it's fine.

(Could be a pathological situation where there is no free core associated with the numa domain. But that's a sign that the affinity settings are messed up.)